### PR TITLE
VTCCA-2975 switching property linking to upscan v2 with needed clean …

### PR DESCRIPTION
**Pull request by DanielConnelly**
…up in certain areas to allow passing tests and to use libraries that simplify upscan v2 integration, the javascript now loads the response from the initiate into the page and submits the html form which redirects to the same page

### CODE REVIEW

File: app/controllers/propertyLinking/UploadController.scala [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+      case EvidenceChoices.OTHER =>
+        form.bindFromRequest().fold(
+          _ =>
+            Future.successful(BadRequest(uploadEvidence(request.ses.submissionId, List("error.businessRatesAttachment.evidence.not.selected"), request.ses.uploadEvidenceData.attachments.getOrElse(Map()), form))),
+          formData => {
+            val updatedSession = session.copy(evidenceType = EvidenceType.fromName(formData.name))
+            val sessionUploadData: UploadEvidenceData = updatedSession.uploadEvidenceData
+              .copy(
+                linkBasis = OtherEvidenceFlag,
+                fileInfo = updatedSession.uploadEvidenceData.fileInfo.map(_.copy(evidenceType = EvidenceType.fromName(formData.name).get))) //.get is dangerous,
```
Comment (PeterPerhac) [inline app/controllers/propertyLinking/UploadController.scala]: agree, `.get` is asking for trouble. basically, if you can't bind a valid EvidenceType from request, then user should go to BadRequest page. at this point you should have a valid EvidenceType. Resolving it from String here leads to having to do .get which can explode

File: app/uk/gov/voa/propertylinking/errorhandler/CustomErrorHandler.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  private def getDateTime: LocalDateTime = {
+    val instant = Instant.ofEpochMilli(System.currentTimeMillis)
+    LocalDateTime.ofInstant(instant, ZoneId.of("Europe/London"))
+  }
+
+  private def extractErrorReference(request: Request[_]): Option[String] = {
+    request.headers.get(HeaderNames.xRequestId) map {
+      _.split("-")(2)
```
Comment (PeterPerhac) [inline app/uk/gov/voa/propertylinking/errorhandler/CustomErrorHandler.scala]: where did this come from? is copy-paste from some other service? doesn't seem safe to do this at all...  you don't know if there are 2 hyphens in that header value. `split` you could wrap the split...(2) in a Try and then .toOption it in a `flatMap`

```scala
request.headers.get(HeaderNames.xRequestId).flatMap ( v => Try(v.split("-")(2)).toOption)
```

File: app/assets/javascripts/src/fileUpload.js [hunk 3]
```diff
@@ -150,27 +151,13 @@
         function fileUpload(form, file, csrfToken){
             var data = new FormData();
```
Comment (reddy-yattapu) [inline app/assets/javascripts/src/fileUpload.js]: var data = new FormData(); is not been used any where else, just remove it

File: app/controllers/propertyLinking/ClaimProperty.scala [hunk 4]
```diff
@@ -73,13 +80,13 @@ class ClaimProperty @Inject()(val authenticated: AuthenticatedAction,
     }
   }
 
-  def declareCapacity(uarn: Long, address: String) = authenticated.async { implicit request =>
+  def declareCapacity(uarn: Long, address: String) = authenticatedAction { implicit request =>
     Ok(views.html.propertyLinking.declareCapacity(DeclareCapacityVM(declareCapacityForm, address, uarn)))
   }
 
-  def attemptLink(uarn: Long, address: String) = authenticated.async { implicit request =>
+  def attemptLink(uarn: Long, address: String): Action[AnyContent] = authenticatedAction.async { implicit request =>
```
Comment (darrell-plant) [inline app/controllers/propertyLinking/ClaimProperty.scala]: do you need .async here?

File: app/controllers/propertyLinking/ClaimProperty.scala [hunk 5]
```diff
@@ -88,12 +95,12 @@ class ClaimProperty @Inject()(val authenticated: AuthenticatedAction,
     )
   }
 
-  def back = withLinkingSession { implicit request =>
+  def back: Action[AnyContent] = authenticatedAction.andThen(withLinkingSession).async { implicit request =>
```
Comment (darrell-plant) [inline app/controllers/propertyLinking/ClaimProperty.scala]: .async?

File: app/uk/gov/voa/propertylinking/errorhandler/CustomErrorHandler.scala [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    views.html.errors.error(pageTitle, heading, message)
+
+  override def internalServerErrorTemplate(implicit request: Request[_]): Html = {
+    views.html.errors.technicalDifficulties(extractErrorReference(request), getDateTime)
+  }
+
+  private def getDateTime: LocalDateTime = {
+    val instant = Instant.ofEpochMilli(System.currentTimeMillis)
+    LocalDateTime.ofInstant(instant, ZoneId.of("Europe/London"))
+  }
```
Comment (darrell-plant) [inline app/uk/gov/voa/propertylinking/errorhandler/CustomErrorHandler.scala]: is this different from LocalDateTime.now ?

Comment (PeterPerhac) [reply app/uk/gov/voa/propertylinking/errorhandler/CustomErrorHandler.scala]: seems copy-paste from one of these places : https://github.com/search?q=org%3Ahmrc+%22++++val+instant+%3D+Instant.ofEpochMilli%28System.currentTimeMillis%29%22&type=Code

File: app/auth/VoaAction.scala [hunk 7]
```diff
@@ -16,20 +16,16 @@
...
 import play.api.Logger
 import play.api.mvc.{Action, AnyContent, Request, Result}
-import uk.gov.hmrc.http.{BadRequestException, HeaderCarrier, NotFoundException}
+import uk.gov.hmrc.http.HeaderCarrier
 import uk.gov.hmrc.play.HeaderCarrierConverter
-import uk.gov.hmrc.play.frontend.auth.{Actions, AuthContext}
-import uk.gov.hmrc.play.frontend.auth.connectors.AuthConnector
 
-import scala.concurrent.{ExecutionContext, Future}
 import scala.concurrent.ExecutionContext.Implicits.global
```
Comment (darrell-plant) [inline app/auth/VoaAction.scala]: Is it okay to use ExecutionContext.Implicits.global?

Comment (PeterPerhac) [reply app/auth/VoaAction.scala]: don't worry, this i am fixing in my current work

File: app/controllers/propertyLinking/ClaimProperty.scala (hunk 4 above)
Comment (DanielConnelly) [reply app/controllers/propertyLinking/ClaimProperty.scala]: coulda used the multi line tool here darrell.

File: app/uk/gov/voa/propertylinking/errorhandler/CustomErrorHandler.scala (hunk 6 above)
Comment (DanielConnelly) [reply app/uk/gov/voa/propertylinking/errorhandler/CustomErrorHandler.scala]: This was copy and paste, I can change to LocalDateTime.now()

File: app/auth/VoaAction.scala (hunk 7 above)
Comment (DanielConnelly) [reply app/auth/VoaAction.scala]: no good catch.

File: app/controllers/propertyLinking/ClaimProperty.scala (hunk 4 above)
Comment (DanielConnelly) [reply app/controllers/propertyLinking/ClaimProperty.scala]: needed here :D

File: app/uk/gov/voa/propertylinking/errorhandler/CustomErrorHandler.scala (hunk 2 above)
Comment (DanielConnelly) [reply app/uk/gov/voa/propertylinking/errorhandler/CustomErrorHandler.scala]: request-id is wanted to be displayed on the screen this is really BROKEN but I copied it from the other place it sat we need to discuss the best way to get the request-id out of the headers for all frontend services.

**Comment by darrell-plant** [comment]
LGTM 👍

**Comment by habeeb01** [comment]
LGTM 👍

**Pull request merged**